Add cookie based routing for selected paths (#188)
trinodb/trino-gateway: docs/routing-logic.md

**Trino Gateway documentation**

<table>
  <tr>
    <td>
      <img src="./assets/logos/trino-gateway-v.png"/>
    </td>
    <td>
      <ul>
        <li><a href="quickstart.md">Quickstart</a></li>
        <li><a href="installation.md">Installation</a></li>
        <li><a href="security.md">Security</a></li>
        <li><a href="operation.md">Operation</a></li>
      </ul>
    </td>
    <td>
      <ul>
        <li><a href="gateway-api.md">Gateway API</a></li>
        <li><a href="resource-groups-api.md">Resource groups API</a></li>
        <li><a href="routing-rules.md">Routing rules</a></li>
        <li><a href="routing-logic.md">Routing logic</a></li>
      </ul>
    </td>
    <td>
      <ul>
        <li><a href="design.md">Design</a></li>
        <li><a href="development.md">Development</a></li>
        <li><a href="release-notes.md">Release notes</a></li>
        <li><a href="references.md">References</a></li>
      </ul>
    </td>
  </tr>
</table>

# Routing Logic

> ## Overview

Reviewer: remove redundant title or add intro sentence above

> Incoming requests are evaluated by the Gateway to determine if they

Reviewer: ```suggestion
Incoming requests are evaluated by Trino Gateway to determine if they
```

Reviewer: and everywhere else .. full name

are a follow-up to a request previously handled by the Gateway. If the 

> Gateway determines that they are, they are forwarded to whatever the 

Reviewer: ```suggestion
Gateway determines that a request is a follow up to a previous request, it is forwarded to the same backend as before. 
```

backend that handled the previous request.

> If the request does not have an antecedent, the Gateway first uses [Routing Rules](routing-rules.md)

Reviewer: ```suggestion
If the request is new, Trino Gateway first uses [Routing rules](routing-rules.md)
```

to determine which logical group of clusters, termed a "Routing Group",
should handle the request. A member of the Routing Group, termed a backend, 
is selected to handle the request using either an adaptive or round-robin 
strategy.

> ![Request Routing Flow](assets/gateway-routing-flow.svg)

Reviewer 2: Can you add a start/user in the flow? I assume it's "is request descended from a previous request"

Author: Not sure why I'm unable to respond to these individually

> Why would you need user to configure this? Other than logout?
There is no generic reason. However, the cookie routing might be leveraged in use case specific ways. For example, maybe some workflow using [Dune's table API](https://docs.dune.com/api-reference/tables/endpoint/create) would benefit from pinning to a specific cluster, then unpinning after a certain request.

Are you looking for more specific guidance on this in the docs?

> In which cases would you need custom rules?
Again, no generic reason, but there could be a use case specific reason. For example, Starburst has a few different APIs for managing access control, caching and data virtualization. There may be some case where you want to unpin depending on something in the response instead of just the path. However, this is out of scope for the current work so I've removed the TODO

> ## Sticky Routing

Reviewer: ```suggestion
## Sticky routing
```

> A request related to an ongoing process or state maintained on a single 

Reviewer: empty line after title here and elsewhere

> A request related to an ongoing process or state maintained on a single 

Reviewer: ```suggestion
A request related to an ongoing process or state, maintained on a single 
```

> backend cluster must be routed to that backend for proper handling. Two 

Reviewer: ```suggestion
backend cluster, must be routed to that backend for proper handling. Two 
```

> mechanisms for identifying related requests are currently implemented.

Reviewer: ```suggestion
mechanisms for identifying related requests are implemented and detailed in the following sections.
```

> ### Query ID Based Routing

Reviewer: Routing based on query identifier

Reviewer 2: How about adding `(default)` at the end?

When a query is initiated through the Gateway, the query id will be 
extracted from the response and mapped to the backend that provided the 
response. Any subsequent request containing that query id will be forwarded 
to that backend. For example, to retrieve query results, the trino client 
polls a URI of the form 
`v1/statement/executing/queryid/nonce/counter`. The Gateway will extract
the queryid from this URI.

> ### Cookie Based Routing

Reviewer: Routing based on cookies

> The Gateway can add a session cookie to identify related requests. This is 

Reviewer: ```suggestion
Trino Gateway can add a session cookie to new request and use it to identify related requests. This behavior is 
```

> useful for exchanges such as the OAuth handshake, where it isn't possible 

Reviewer: Change this to talk about it being necessary for Trino cluster that use OAuth authentication ... so that it hits the same cluster based on the limited info

to determine which backend handled the previous request based solely on
the request URI and body.

> #### Cookie Based Routing Configuration

Reviewer: Drop that title and replace with sentence about configuring .. adding where to configure this .. .. and say .. with the following properties

maybe have an example snippet

Reviewer 2: Please do add examples :)

> `cookiePaths`: If the request URI starts with a path in this list, add a 

Reviewer: list? comma separated? add example

session cookie if none exists. If a cookie exists, route the request to
the backend associated with that cookie.

> `removeCookiePaths`: If the request URI starts with a path in this list,

Reviewer: why do you need that? In what situation should a user configure that .. to what value

Reviewer 2: Why would you need user to configure this? Other than logout?

return a response that instructs the client to delete the cookie.

#### Cookie Based Routing for OAuth2
The default OAuth2 handshake for the Trino REST API has the following steps 

> 1. The client requests https://trino/v1/endpoint. The response includes

Reviewer: empty line before list

```
HTTP/2 401
www-authenticate: Bearer x_redirect_server="https://trino/oauth2/token/initiate
, x_token_server="https://trino/oauth2/token/encoded_token"
```

> which prompts the client to use oauth2 authentication.

Reviewer: ```suggestion
which prompts the client to use OAuth2 authentication.
```

> 2. GET https://trino/oauth2/token/encoded_token. The response redirects to the IdP's auth endpoint.

Reviewer: need to spell out IdP since this is first usage on this page

> 3. GET https://idp/auth_endpoint. The response includes an authorization code and 

Reviewer: better example and explain where that comes from

redirects the client back to the coordinator's callback endpoint.
4. GET https://trino/oauth2/callback?authroization_code_parameters. The coordinator 
sends the authorization code to the IdP in combination with secrets generated in 
step (1). The IdP returns an access token to the coordinator, which encodes it and sends
it to the client.

> Adding the configuration

Reviewer: Use the following configuration (where) to include...

```yaml
cookiePaths: /oauth2
```
will cause the Gateway to include a session cookie in the response to the

> request to `/oauth2/token` endpoint. This session cookie will be carried through the client's request

Reviewer: ```suggestion
request to `/oauth2/token` endpoint. The session cookie is carried through the client's request
```

> to the IdP auth endpoint and `/oauth2/callback`. The session cookie ensures that the request

Reviewer: ```suggestion
to the IdP authentication endpoint and `/oauth2/callback`. The session cookie ensures that the request
```

to `/oauth2/callback` is made to the backend cluster that initiated the exchange, which
is the only backend which has the secrets that allow it to use the authorization code. 

The handshake for the Web UI is slightly different, but because it still uses endpoints
under `/oauth2`, the same configuration supports a Web UI secured with OAuth2. Adding

> ```yaml

Reviewer: Add the following configuration to force removal...

deleteCookiePaths: /ui/logout
```
will force removal of the session cookie when a user logs out.

> This configuration will not affect requests to any `v1/*` or other Trino endpoints, so

Reviewer: Which configuration ? The cookie-based routing configuration?

```suggestion
The cookie  configuration does not affect requests to any `v1/*` or other Trino endpoints.
```

> queries and other requests will still be routed according to Routing Rules.

Reviewer: Drop
